motor-controller/common/pid.c

```diff
@@ -45,6 +41,142 @@ float pid_calculate(Pid_t *pid, float r, float y, float dt) {
+        pid->cur_gain_stage_ind = 0;
+    } else if (y < pid->gain_schedule[pid->num_gain_stages - 1]) {
+        // we are above the highest gain stage, not between
+        pid->cur_gain_stage_ind = pid->num_gain_stages - 1;
+    } else {
+        // we are between gain stages
+
+        for (size_t i = 0; i < pid->num_gain_stages - 1; i++) {
+            // find the gain stages we are between
+
+            size_t lower_gain_stage_ind = i;
+            size_t upper_gain_stage_ind = i + 1;
+            float lower_gain_stage_point = pid->gain_schedule[lower_gain_stage_ind];
+            float upper_gain_stage_point = pid->gain_schedule[upper_gain_stage_ind]; 
+
+            if (lower_gain_stage_point <= y && y < upper_gain_stage_point) {
+                // we found the gain stages 
+
+                float midpoint = (lower_gain_stage_point + upper_gain_stage_point) / 2.0f;
+                float stage_range_hyst = (upper_gain_stage_point - lower_gain_stage_point) * pid->hyst;
+                if (pid->cur_gain_stage_ind == lower_gain_stage_ind) {
+                    // the current gain index is the lower value, apply hyst to upper thresh
+
+                    if (y > midpoint + stage_range_hyst) {
+                        pid->cur_gain_stage_ind = upper_gain_stage_ind;
+                    } else {
+                        // stay where we are
+                    }
+                } else if (pid->cur_gain_stage_ind == upper_gain_stage_ind) {
+                    // the current gain index is the upper value, apply hist the lower thresh
+
+                    if (y < midpoint - stage_range_hyst) {
+                        pid->cur_gain_stage_ind = lower_gain_stage_ind;
+                    } else {
+                        // stay where we are
+                    }
+                } else {
+                    // the current gain index is neither, the system state has moved rapidly, choose the closest
+                    if (y < midpoint) {
+                        pid->cur_gain_stage_ind = lower_gain_stage_ind;
+                    } else {
+                        pid->cur_gain_stage_ind = upper_gain_stage_ind;
+                    }
+                }
+            }
+        }
+    }
+}
+
+float gspid_calculate(GainScheduledPid_t *pid, float r, float y, float dt) {
+    // choose correct gains for current state
+    gspid_update_gain_stage(pid, y);
+    PidConstants_t cur_gains = pid->pid_constants[pid->cur_gain_stage_ind];
+
+    float err = r - y;
+
+    float termP = err * cur_gains.kP;
+
+    pid->eI = pid->eI + (err * dt);
+
+    if (pid->eI > cur_gains.kI_max) {
+        pid->eI = cur_gains.kI_max;
+    } else if (pid->eI < cur_gains.kI_min) {
+        pid->eI = cur_gains.kI_min;
+    }
+    float termI = pid->eI * cur_gains.kI;
+
+    float termD = ((err - pid->prev_err) / dt) * cur_gains.kD; // flip err and prev_err???
+    pid->prev_err = err;
+
     float u = r + (termP + termI + termD);
     return u;
```

Comment:
It'd be nice (but obviously not required) to make the core implementation of PID shared between the scheduled and non-scheduled functions.